src/runtime/context.ts

```diff
@@ -22,7 +23,7 @@ export class Context {
   evaluatedRecords: any[];
 
   constructor(
-    parent: Context,
+    parent: Context | Library,
```

Comment:
I agree; some of the stuff around `Context` is confusing.  I feel like this is an area that could benefit from some refactoring.

> So a context can either be a "root" context, or the child of some other context... Is the parent of a "root context" just the library itself?

I _think_ that the root context is typically an instance of `PatientContext` or `UnfilteredContext`.  And the "parent" of `PatientContext` and `UnfilteredContext` is always a library.  If you look in their constructors, they both call `super(library, ...)` (and that first argument is the parent). When I changed their `library` properties from `any` to `Library`, TypeScript got mad; which is why I changed `Context.parent` to allow `Library`. (TBH, I was a little surprised that TS had no problem with me passing an `any` where `Context` is expected; I guess I don't understand TS as well as I thought).

> If that is the case, is it theoretically unsafe to access things like this.parent.codeService or this.parent.rootContext, etc. which don't exist on type Library?

That's a good question.  Maybe?  Although I'm not aware of it ever actually causing a problem, so perhaps it is just theoretical...  OK... I guess I can't just let this one go, so let's take a look:
* `this.parent.rootContext()`: Probably OK. I don't think that `this.parent.rootContext` would ever get called when a `Library` is the parent because `PatientContext` and `UnfilteredContext` are the only ones with a `Library` as parent, and they both override `rootContext` to always return `this`.
* `this.parent.codeService`: Probably OK, because it's only called if the context doesn't have a `_codeService`. In which case, neither does `Library`, so `undefined` is a reasonable answer to give.
* `this.parent.parameters`: OK. It's only called if the context doesn't have a `_parameters`, but `Context`'s `_parameters` defaults to `{}` when not passed in -- so it's unlikely `_parameters` would ever be undefined. But if it was, `Library` has `parameters` too -- and it is initialized to `{}` even when there are no parameters.
* `this.parent.findRecords(profile, retrieveDetails)`: OK, because `PatientContext`/`UnfilteredContext` override `findRecords` without invoking the `parent`.
* `this.parent.getLibraryContext(library)`: OK, because `PatientContext`/`UnfilteredContext` override `getLibraryContext` with invoking the `parent`.
* `this.parent.getLocalIdContext(localId)`: Not sure. `PatientContext` overrides `getLocalIdContext()`, but `UnfilteredContext` does not.  So there might be an issue if someone tried to `getLocalIdContext(localId)` on an `UnfilteredContext`. It looks like `getLocalIdContext` is not called anywhere in `cql-execution` -- so I'm guessing this was added for measure highlighting an such; and since measures weren't using `UnfilteredContext`, it didn't matter...
* `this.parent.getParameter(name)`: OK, because `Library` implements `getParameter`.
* `this.parent.getParentParameter(name)`: Not sure. `Library` does not define `getParentParameter`.  But... `this.parent.getParentParameter(name)` is only called if the parameter isn't first found in the `this.parent.parameters` object.  I think it will always be found there first, because the only way you could possibly look for a parameter that isn't there is if your CQL is invalid (i.e., a library that tries to access a parameter that doesn't exist won't compile to ELM in the first place).
* `this.parent.getTimezoneOffset()`: OK. This only gets called if `this.parent.getTimezoneOffset != null` -- so there's already a guard in place, maybe for this very reason!
* `this.parent.getExecutionDateTime()`: OK. This only gets called if `this.parent.getExecutionDateTime != null`.
* `this.parent.getMessageListener()`: OK. This only gets called if `this.parent.getMessageListener != null`.
* `this.parent.getValueSet(name, library)`: OK, because `Library` implements `getValueSet`.
* `this.parent.getCodeSystem(name)`: OK, because `Library` implements `getCodeSystem`.
* `this.parent.getCode(name)`: OK, because `Library` implements `getCode`.
* `this.parent.getConcept(name)`: OK, because `Library` implements `getConcept`.
* `this.parent.get(identifier)`: OK, because `Library` implements `get`. `UnfilteredContext` also overrides `get` without invoking the `parent`.
* `this.parent.source.library.identifier.id`: OK, because `Library` has `source` set to the JSON.

So... I think the only questionable ones are `getLocalIdContext(localId)` and `getParentParameter(name)`.  What do you think? Should I try to do something about those in this PR?

Reply:
If we want to do something, I think we have two choices:
1. Just fix the specific things about `getLocalIdContext(localId)` and `getParentParameter(name)` that are problematic.
2. Make `Library` extend `Context` and then override any of the function implementations that don't make sense for `Library`. This would then also allow me to revert the `Context | Library` change.

Reply:
Got it. Thanks for the explanation. I think it probably makes more sense to pawn off some refactoring of context and the types associated with it to separate work on this repo, and merge this PR as is. Thought?

Reply:
Sure. That seems fair. I just filed #283 for that.